File: backend-dotnet/src/Core/App.Template.Domain/Entities/AuditLog.cs
namespace AppTemplate.Domain.Entities;

/// <summary>
/// Represents an audit log entry that tracks changes to entities.
/// </summary>
public class AuditLog
{
    public long Id { get; private set; }

    /// <summary>
    /// Name of the entity that was modified (e.g., "User", "Department")
    /// </summary>
    public string EntityName { get; private set; } = string.Empty;

    /// <summary>
    /// Primary key of the entity that was modified
    /// </summary>
    public string EntityId { get; private set; } = string.Empty;

    /// <summary>
    /// The type of action performed: Created, Updated, Deleted
    /// </summary>
    public AuditAction Action { get; private set; }

    /// <summary>
    /// JSON representation of the entity's values before the change (null for Created)
    /// </summary>
    public string? OldValues { get; private set; }

    /// <summary>
    /// JSON representation of the entity's values after the change (null for Deleted)
    /// </summary>
    public string? NewValues { get; private set; }

    /// <summary>
    /// List of property names that were modified (JSON array)
    /// </summary>
    public string? AffectedColumns { get; private set; }

    /// <summary>
    /// User ID of who performed the action
    /// </summary>
    public string? UserId { get; private set; }

    /// <summary>
    /// UTC timestamp when the action occurred
    /// </summary>
    public DateTime Timestamp { get; private set; }

    private AuditLog() { }

    public static AuditLog Create(
        string entityName,
        string entityId,
        AuditAction action,
        string? oldValues,
        string? newValues,
        string? affectedColumns,
        string? userId)
    {
        return new AuditLog
        {
            EntityName = entityName,
            EntityId = entityId,
            Action = action,
            OldValues = oldValues,
            NewValues = newValues,
            AffectedColumns = affectedColumns,
            UserId = userId,
            Timestamp = DateTime.UtcNow
        };
    }
}

public enum AuditAction
{
    Created,
    Updated,
    Deleted
}

File: backend/spring/clean-architecture/full/domain/src/main/java/apptemplate/domain/entities/AuditLog.java
package apptemplate.domain.entities;

import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.Builder;

import java.time.LocalDateTime;

/**
 * Represents an audit log entry that tracks changes to entities.
 */
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class AuditLog {

    private Long id;

    /**
     * Name of the entity that was modified (e.g., "User", "Department")
     */
    private String entityName;

    /**
     * Primary key of the entity that was modified
     */
    private String entityId;

    /**
     * The type of action performed: CREATED, UPDATED, DELETED
     */
    private AuditAction action;

    /**
     * JSON representation of the entity's values before the change (null for Created)
     */
    private String oldValues;

    /**
     * JSON representation of the entity's values after the change (null for Deleted)
     */
    private String newValues;

    /**
     * List of property names that were modified (JSON array)
     */
    private String affectedColumns;

    /**
     * User ID of who performed the action
     */
    private Long userId;

    /**
     * UTC timestamp when the action occurred
     */
    private LocalDateTime timestamp;

    public enum AuditAction {
        CREATED,
        UPDATED,
        DELETED
    }
}
